x/shield/keeper/proposal.go

```diff
@@ -11,12 +11,26 @@ import (
 )
 
 // ClaimLock locks collaterals after a claim proposal is submitted.
-func (k Keeper) ClaimLock(ctx sdk.Context, proposalID uint64, poolID uint64,
-	loss sdk.Coins, purchaseTxHash []byte, lockPeriod time.Duration) error {
+func (k Keeper) ClaimLock(ctx sdk.Context, proposalID uint64, poolID uint64, loss sdk.Coins, purchaseTxHash []byte, lockPeriod time.Duration) error {
 	pool, err := k.GetPool(ctx, poolID)
 	if err != nil {
 		return err
 	}
+	if !pool.Shield.IsAllGTE(loss) {
+		panic(types.ErrNotEnoughShield)
+	}
+
+	lossAmt := loss.AmountOf(k.sk.BondDenom(ctx))
+	poolValidCollateral := sdk.ZeroInt()
+	collaterals := k.GetAllPoolCollaterals(ctx, pool)
+	for _, collateral := range collaterals {
+		if collateral.Amount.IsPositive() {
+			poolValidCollateral = poolValidCollateral.Add(collateral.Amount)
```

Comment:
Ideally we want `collateral.Amount` to be always nonnegative, right? Do you know why this could be negative?

Reply:
Also, should we subtract `collateral.TotalLocked` here to determine whether we have enough to lock?

Reply:
`collateral.Amount` can be negative. Because withdraw period is 21 days while claim period is 21 + 4 (voting period) days, it is possible that after A finished withdraw it still needs to pay for the loss. However, A will not be able to get real tokens because we track A's unbonding delegations.

Reply:
A negative collateral means `A` finished some withdraw which it shouldn't do. It will be used to calculate "valid" collaterals.

The reason we don't modify the withdraw time like we did to the unbonding time is that we should not ask `A` to be responsible to the losses 21 days after it started its withdrawal.

Reply:
Claim period is 21 days, and if it is submitted later than 17 days we extend the existing withdraws. It's confusing to say claim period is 25 days.

Reply:
> Claim period is 21 days, and if it is submitted later than 17 days we extend the existing withdraws. It's confusing to say claim period is 25 days.

Okay I should say the last possible day of paying the loss. I prefer to set negative collateral amount than postpone existing withdraws because the latter would be more complicated.

For example, `A` withdrew all its 100ctk collateral on Day 0, but a claim proposal locked 50ctk of its collateral on Day 20. If we postponed the withdraw of this 50ctk and the claim proposal failed the certifier voting round, this 50ctk would be unlocked and be responsible to losses in another 4 days. The solution is keeping a record of the original withdraw date and set the original withdraw date when the collateral is unlocked, but it will bring some extra work. Therefore, I would prefer to keep this design unless it is incorrect.

Reply:
Yeap. Just extending won't do the job. I agree

Reply:
Makes sense. Thanks for the explanation.